---
title: "5xBOAT_snRNAseq_Subclustering"
author: "Joshua Samuels"
date: "11/3/2022"
output: html_document
---


#### Load Packages

```{r, message=FALSE, warning=FALSE}
# Load Packages
library(Seurat)
library(tidyverse)
library(Matrix)
library(scales)
library(cowplot)
library(RCurl)
library(ggridges)
library(BiocManager)
library(ggplot2)
library(dplyr)
library(magrittr)
library(patchwork)
library(ggraph)
library(clustree)
library(metap)
library(limma)
```


#### Load labeled clustered Seurat of interest
```{r}
subset_seurat <-readRDS(
  "D:/Research/UVA_NGP_PhD/Lukens_Lab/Experiments/BOAT_MG_KO/Phenotype_Exploration/snRNAseq/RStudio_SeuratObjects/MG_seurat_11.3.22.rds")

DefaultAssay(subset_seurat) <- "RNA"
subset_seurat <- DietSeurat(subset_seurat, assays = "RNA")
subset_seurat@meta.data[["integrated_snn_res.0.1"]] <- NULL
subset_seurat@meta.data[["integrated_snn_res.0.2"]] <- NULL
subset_seurat@meta.data[["integrated_snn_res.0.3"]] <- NULL
subset_seurat@meta.data[["integrated_snn_res.0.4"]] <- NULL
subset_seurat@meta.data[["integrated_snn_res.0.8"]] <- NULL
subset_seurat@meta.data[["seurat_clusters"]] <- NULL
```


#### Redo SCTransform, PCA, UMAP, and Clustering
This method is based off https://github.com/satijalab/seurat/issues/1883. 

##### Re-SCTransform
```{r}
# load the sctransfrom package
library(sctransform)

# split seurat object by sample
subset.sct.list <-SplitObject(subset_seurat, split.by = "sample")

# normalize and identify variable features for each seurat object independently
subset.sct.list <- lapply(X = subset.sct.list, FUN = SCTransform)

# select features that are repeatedly variable across seurat objects for integration
subset.sct.features <- SelectIntegrationFeatures(object.list = subset.sct.list, nfeatures = 3000)

# pep for integration steps
subset.sct.list <- PrepSCTIntegration(object.list = subset.sct.list, anchor.features = subset.sct.features)

# identifies anchors for integration - takes a long time!
subset.sct.anchors <- FindIntegrationAnchors(object.list = subset.sct.list, 
                                      normalization.method = "SCT", 
                                      anchor.features = subset.sct.features)

# perform integration
sct_subset_seurat <- IntegrateData(anchorset = subset.sct.anchors, normalization.method = "SCT")
```

```{r}
# Remove SCTransform prep objects from the environment
rm(subset_seurat)
rm(subset.sct.anchors)
rm(subset.sct.list)
rm(subset.sct.features)
```


##### Re-PCA
```{r}
# Rename seurat
subset_seurat_integrated <- sct_subset_seurat

# remove unused data from environment
rm(sct_subset_seurat)

# use integrated data to assess dimensionality and resolution
DefaultAssay(subset_seurat_integrated) <- "integrated"

### PCA
# Run PCA
subset_seurat_integrated <- RunPCA(object = subset_seurat_integrated, verbose = FALSE)

### Check dimensionality
ElbowPlot(subset_seurat_integrated, ndims = 50)

### UMAP
# Run UMAP
# I am using a very high dimensionality since SCTransform allows for higher dimensionality - see Seurat "Using sctransform in Seurat" vignette - https://satijalab.org/seurat/articles/sctransform_vignette.html
subset_seurat_integrated <- RunUMAP(subset_seurat_integrated, reduction = "pca", dims = 1:15, verbose = FALSE)
```


##### Re-Clustering
```{r}
### Clustering
# Determine the K-nearest neighbor graph
subset_seurat_integrated <- FindNeighbors(object = subset_seurat_integrated, dims = 1:15, verbose = FALSE)

# Determine the clusters for various resolutions                                
subset_seurat_integrated <- FindClusters(object = subset_seurat_integrated,
                               resolution = c(0.1, 0.15, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8))

# Plot UMAPs for various clustering resolutions
# Assign identity of clusters based on resolution
Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.1"
# Plot the UMAP
res.1 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        split.by = "sample",
        label.size = 6)
res.1

Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.2"
# Plot the UMAP
res.2 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        split.by = "sample",
        label.size = 6)
res.2

Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.3"
# Plot the UMAP
res.3 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
res.3

Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.4"
# Plot the UMAP
res.4 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
res.4

Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.5"
# Plot the UMAP
res.5 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
res.5

# Assign identity of clusters based on resolution
Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.6"
# Plot the UMAP
res.6 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
res.6

# Assign identity of clusters based on resolution
Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.7"
# Plot the UMAP
res.7 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
res.7

# Assign identity of clusters based on resolution
Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.8"
# Plot the UMAP
res.8 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
res.8
```

```{r}
# Choose a resolution value! 
# Assign identity of clusters based on resolution of choice
Idents(object = subset_seurat_integrated) <- "integrated_snn_res.0.1"
# Plot the UMAP
res.1 <- DimPlot(subset_seurat_integrated,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
res.1

# change name of seurat object to clustered_seurat
subset_seurat_clustered <- subset_seurat_integrated
```

```{r}
# remove unused resolution plots from the environment
rm(res.2)
rm(res.3)
rm(res.4)
rm(res.5)
rm(res.6)
rm(res.7)
rm(res.8)
rm(subset_seurat_integrated)
```


##### Observe final Re-clustered UMAP
```{r}
# final clustered UMAP (should match the UMAP with the chosen resolution above)
p4 <- DimPlot(subset_seurat_clustered, 
              reduction = "umap", repel = T, label.box = T, pt.size = .8, cols = c(
        "skyblue1",
        "lightcoral",
        "palegreen",
        "violet"))
p4
# final clustered UMAP grouped by sample (able to observe overlap to suggest good integration)
p5 <- DimPlot(subset_seurat_clustered, reduction = "umap", group.by = "sample")
# final clustered UMAP split by sample
p6 <- DimPlot(subset_seurat_clustered, reduction = "umap", split.by = "sample", pt.size = .8, cols = c(
        "skyblue1",
        "lightcoral",
        "palegreen",
        "violet"))

LabelClusters(plot=p4, id= "ident", repel = T) + NoLegend()
p5
LabelClusters(plot=p6, id= "ident", repel = T) + NoLegend()
```

```{r}
# plot p4 (numbered UMAP)
LabelClusters(plot=p4, id= "ident", repel = T) + NoLegend()

# Save last plot that was printed
# Before saving, set your size and dpi (quality)
# ggsave("FileName.FileType", path = "path/path", width = , height = , units = "cm", dpi = 500)

ggsave("MGsubset_seurat_umap_numbered_11.15.22.tiff", 
       path = "D:/Research/UVA_NGP_PhD/Lukens_Lab/Experiments/BOAT_MG_KO/Phenotype_Exploration/snRNAseq/RStudio_SeuratDataFiles/UMAPs/", 
       height = 5, 
       width = 7, 
       dpi = 500)
```

```{r}
# plot p6 (numbered UMAP, split by sample)
LabelClusters(plot=p6, id= "ident", repel = T) + NoLegend()

# Save last plot that was printed
# Before saving, set your size and dpi (quality)
# ggsave("FileName.FileType", path = "path/path", width = , height = , units = "cm", dpi = 500)

ggsave("MGsubset_seurat_umap_numbered_SampleSplit_11.15.22.tiff", 
       path = "D:/Research/UVA_NGP_PhD/Lukens_Lab/Experiments/BOAT_MG_KO/Phenotype_Exploration/snRNAseq/RStudio_SeuratDataFiles/UMAPs/", 
       height = 5, 
       width = 10, 
       dpi = 500)
```


#### Extract and save the number of cells per cluster
```{r}
# Extract identity and sample information from seurat object to determine the number of cells per cluster per sample
n_cells <- FetchData(subset_seurat_clustered, 
                     vars = c("ident", "sample")) %>%
        dplyr::count(ident, sample) %>%
        tidyr::spread(ident, n)

# View table
View(n_cells)

# Save the number of cells per cluster as a .csv file
write.csv(n_cells, file = 
            "D:/Research/UVA_NGP_PhD/Lukens_Lab/Experiments/BOAT_MG_KO/Phenotype_Exploration/snRNAseq/RStudio_SeuratDataFiles/nCells_MGsubset_res0.1_11.15.22.csv")
```

```{r}
# remove unused data from the environment
rm(n_cells)
rm(res.1)
```


#### Analyzing Cluster Identities

```{r}
# use RNA data to assess gene expression contributing to clustering
DefaultAssay(subset_seurat_clustered) <- "RNA"

# DefaultAssay(subset_seurat_clustered) <- "SCT"
# subset_seurat_clustered <- PrepSCTFindMarkers(subset_seurat_clustered)
```


##### Using canonical markers
```{r}
# identify canonical cell type marker genes that are conserved across conditions
# will need to do this for individual clusters

# Cluster 0
all.conserved.markers.0 <- FindConservedMarkers(subset_seurat_clustered, ident.1 = 0, grouping.var = "sample", verbose = FALSE)
```

##### FindMarkers between clusters directly
```{r}
# Able to put in cluster numbers to the ident.1 and ident.2 functions to compare specific clusters. Able to put in multiple cluster numbers into ident.2 to compare ident.1 against.
cluster.markers <- FindMarkers(subset_seurat_clustered,
                          ident.1 = c(0),
                          ident.2 = c(1),
            only.pos = F,
            # min.diff.pct = 0.25,
            # min.pct = 0.20,
            logfc.threshold = 0.1)

# Converts rownames to 1st column with the column name as "Gene"
cluster.markers <- tibble::rownames_to_column(cluster.markers, "Gene")
view(cluster.markers)

# Export Sig markers to .csv file. Make sure to save file as .xls file!
write.csv(cluster.markers, file = 
            "D:/Research/UVA_NGP_PhD/Lukens_Lab/Experiments/BOAT_MG_KO/Phenotype_Exploration/snRNAseq/RStudio_SeuratDataFiles/MGsubset0v2_res0.1_11.16.22.csv")
```



##### Using FindAllMarkers
```{r}
# Find all sig markers for each cluster compared to all other clusters. Save list to ".csv" file.
subset_All_Sig_Markers <- FindAllMarkers(
  subset_seurat_clustered, only.pos = TRUE, assay = "RNA", logfc.threshold = 0.1)

# Prepare the list of Sig markers
subset_All_Sig_Markers_List <- subset_All_Sig_Markers %>%
  group_by(cluster) %>%
  top_n(n = 100, wt = avg_log2FC)

# Export Sig markers to .csv file. Make sure to save file as .xls file!
write.csv(subset_All_Sig_Markers_List, file = 
            "D:/Research/UVA_NGP_PhD/Lukens_Lab/Experiments/BOAT_MG_KO/Phenotype_Exploration/snRNAseq/RStudio_SeuratDataFiles/MGsubset_AllSigMarkers_res0.1_11.15.22.csv")
```

```{r}
# Remove unused data from the environment
rm(subset_All_Sig_Markers)
rm(subset_All_Sig_Markers_List)
```


###### Test Gene Expression Across Clusters
```{r}
# use a feature plot to show genes of interest
FeaturePlot(subset_seurat_clustered,
            reduction = "umap", 
            features = c("nUMI", "nGene", "mitoRatio", "log10GenesPerUMI"),
            # split.by = "sample",
            order = TRUE,
            # min.cutoff = 'q10',
            label = TRUE,
            cols = c("lightgrey", "red"),
            pt.size = 2)
```

```{r}
# Create stacked violin plot

# Used code from this place https://divingintogeneticsandgenomics.rbind.io/post/stacked-violin-plot-for-visualizing-single-cell-data-in-seurat/
# remove the x-axis text and tick
# plot.margin to adjust the white space between each plot.
# ... pass any arguments to VlnPlot in Seurat
modify_vlnplot<- function(obj, 
                          feature, 
                          pt.size = 0, 
                          plot.margin = unit(c(-0.75, 0, -0.75, 0), "cm"),
                          ...) {
  p<- VlnPlot(obj, features = feature, pt.size = pt.size, ... )  + 
    xlab("") + ylab(feature) + ggtitle("") + 
    theme(legend.position = "none", 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(), 
          axis.title.y = element_text(size = rel(1), angle = 0), 
          axis.text.y = element_text(size = rel(1)), 
          plot.margin = plot.margin ) 
  return(p)
}

## extract the max value of the y axis
extract_max<- function(p){
  ymax<- max(ggplot_build(p)$layout$panel_scales_y[[1]]$range$range)
  return(ceiling(ymax))
}


## main function
StackedVlnPlot<- function(obj, features,
                          pt.size = 0, 
                          plot.margin = unit(c(-0.75, 0, -0.75, 0), "cm"),
                          ...) {
  
  plot_list<- purrr::map(features, function(x) modify_vlnplot(obj = obj,feature = x, ...))
  
  # Add back x-axis title to bottom plot. patchwork is going to support this?
  plot_list[[length(plot_list)]]<- plot_list[[length(plot_list)]] +
    theme(axis.text.x=element_text(), axis.ticks.x = element_line())
  
  # change the y-axis tick to only max value 
  ymaxs<- purrr::map_dbl(plot_list, extract_max)
  plot_list<- purrr::map2(plot_list, ymaxs, function(x,y) x + 
                            scale_y_continuous(breaks = c(y)) + 
                            expand_limits(y = y))

  p<- patchwork::wrap_plots(plotlist = plot_list, ncol = 1)
  return(p)
  
#### Now able to use: StackedVlnPlot() to create a stacked violin plot that is easy to read with multiple features
}
```

```{r}
# Test features
DAM1.Genes <- c("Tyrobp", "Ctsb", "Ctsd", "Apoe", "B2m", "Fth1", "Lyz2")

DAM2.Genes <- c("Trem2", "Axl", "Cst7", "Ctsl", "Lpl", "Cd9", "Cdf1", "Ccl6", "Itgax", "Clec7a", "Lilrb4", "Timp2", "Spp1")

Other.Genes <- c("Inpp5d", "Ptpn6", "Cd74", "H2-d1", "Syk", "Myo1e", "Top2a", "Mki67", "Cd81", "Cd33")

Homeo.Genes <- c("Cx3cr1", "P2ry12", "P2ry13", "Tmem119", "Hexb", "Sall1", "Sall3", "Gpr34", "Sparc", "Itgam", "Tnf", "Cxcr4")
```

```{r}
# use a stacked violin plot to show genes of interest
# Any feature that is not detected will cause an error. Features not found must be commented out and then rerun.

StackedVlnPlot(obj = subset_seurat_clustered, features = c("Hexb", "P2ry12", "Cx3cr1", "C1qa", "Csf1r", "Dock8", "Lrmda"))
```

```{r}
VlnPlot(subset_seurat_clustered, features = "Ctsb", 
          # split.by = "sample", 
          # pt.size = .2, 
          # ncol = 4,
          combine = FALSE
          )

DotPlot(subset_seurat_clustered, features = c(DAM1.Genes, DAM2.Genes, Other.Genes, Homeo.Genes), 
        split.by = "sample",
        cols = c("limegreen", "magenta"),
        assay = "RNA") + RotatedAxis()
```


#### Modify Clusters

##### Rename and Merge Clusters
```{r}
# Assign identity of clusters based on resolution of choice
Idents(object = subset_seurat_clustered) <- "integrated_snn_res.0.1"


# Rename all identities ("old cluster name" = "new cluster name")
# Clusters with the same name will be merged together
subset_seurat_clustered_labeled <- RenameIdents(subset_seurat_clustered, 
                               "0" = "Microglia 1",
                               "1" = "Microglia 2",
                               "2" = "Microglia 3",
                               "3" = "Border-Associated Macropage"
                               )
```

```{r}
# Remove unused data from the environment
rm(subset_seurat_clustered)
```


##### Plot Fully Labeled UMAP
```{r}
# Label the UMAP with your cluster names set above and remove the legend
p7<- DimPlot(object = subset_seurat_clustered_labeled, 
        reduction = "umap", repel = T, pt.size = .8, cols = c(
        "skyblue1",
        "lightcoral",
        "palegreen",
        "violet"))
p7
LabelClusters(plot=p7, id= "ident", box = T, nudge_x = 1, nudge_y = 1) + NoLegend()

```


#### Save subsetted labeled clustered Seurat
```{r}
# Create .RData object to load at any time
saveRDS(subset_seurat_clustered_labeled, file=
          "D:/Research/UVA_NGP_PhD/Lukens_Lab/Experiments/BOAT_MG_KO/Phenotype_Exploration/snRNAseq/RStudio_SeuratObjects/MGsubset_seurat_clustered_labeled_11.15.22.rds")
```


#### Move to next scirpt: 5xBOAT_snRNAseq_DE_Analysis
